news-bundle/src/Controller/NewsFeedController.php

```diff
@@ -0,0 +1,257 @@
+
+/**
+ * @Page(contentComposition=false)
+ */
+class NewsFeedController implements DynamicRouteInterface
+{
+    public const TYPE = 'news_feed';
+
+    private $urlSuffixes = [
+        'atom' => '.xml',
+        'json' => '.json',
+        'rss' => '.xml',
+    ];
+
+    // TODO: Fix content types, maybe make it configurable in the page settings
+    private $contentTypes = [
+        'atom' => 'application/atom+xml',
+        'json' => 'application/feed+json',
+        'rss' => 'application/rss+xml',
+    ];
+
+    public function __construct(private ContaoFramework $framework, private FeedIo $feedIo, private ResponseTagger $responseTagger, private ContaoContext $contaoContext, private ImageFactory $imageFactory, private RequestStack $requestStack, private InsertTagParser $insertTags, private string $projectDir)
+    {
+    }
+
+    public function __invoke(Request $request, PageModel $pageModel): Response
+    {
+        $feed = new Feed();
+        $feed->setTitle($pageModel->title);
+        $feed->setDescription($pageModel->feedDescription);
+        $feed->setLanguage($pageModel->language);
+
+        $time = time();
+        $lastModified = null;
+
+        $archives = StringUtil::deserialize($pageModel->newsArchives);
+        $newsModel = $this->framework->getAdapter(NewsModel::class);
+
+        if ($pageModel->maxFeedItems > 0) {
+            $articles = $newsModel->findPublishedByPids($archives, null, $pageModel->maxFeedItems);
+        } else {
+            $articles = $newsModel->findPublishedByPids($archives);
+        }
+
+        // TODO: Is this still necessary?
+        $origObjPage = $GLOBALS['objPage'] ?? null;
+
+        foreach ($articles as $article) {
+            // Never add unpublished elements to the RSS feeds
+            if (!$article->published || ($article->start && $article->start > $time) || ($article->stop && $article->stop <= $time)) {
+                continue;
+            }
+
+            // TODO: Throw an event here to allow for custom model-to-item transformation?
+            $item = $this->transformModelToItem($article, $pageModel->feedSource, $pageModel);
+
+            if (null === $lastModified) {
+                $lastModified = $item->getLastModified();
+            }
+
+            if ($lastModified < $item->getLastModified()) {
+                $lastModified = $item->getLastModified();
+            }
+
+            $feed->add($item);
+        }
+
+        // TODO: Is this still necessary?
+        $GLOBALS['objPage'] = $origObjPage;
+
+        $feed->setLastModified($lastModified);
+
+        $response = new Response($this->feedIo->format($feed, $pageModel->feedFormat));
+        $response->setLastModified($feed->getLastModified());
+        $response->headers->set('Content-Type', $this->contentTypes[$pageModel->feedFormat]);
+
+        foreach ($archives as $archive) {
+            $this->responseTagger->addTags(['contao.db.news_archive.'.$archive]);
+        }
+
+        return $response;
```

Comment:
should this also be tagged with `tl_page.ID`? or is that done automatically for pages @Toflar ?

Reply:
No tagging is always the task of the respective controller. So it's done automatically for a page regular but not here :) But yes it should be tagged in order to remove the cache entry when you adjust the page alias etc. Should probably also add the jumpTo page ID of the news archive because if that changes, the URLs to the entries might change.

HOWEVER 😊 You are currently not caching anything. You only set the last modified header so the reverse proxy must revalidate and thus the tags are irrelevant ;) But we have cache settings in the page settings, why not applying them here too?

Reply:
@bezin seems like this still applies?

Reply:
I think this was solved here: https://github.com/contao/contao/pull/4682/files#diff-e860c0628b064abb3cac479c7bafd78995fc7a8be5ef97dfc53f06580739fa69R84